### ISSUE TITLE
Disconnected Event Handler not called on reception of CloseMessage

### ISSUE THREAD

**Issue by RobertK66** [Tapanila/SharpCaster#226]
Hi,
my Chromecast (audio speaker) device has an On/Off key.
When used to stop playback on device, the ChromeCastClient receives a "CloseMessage" on the ConnectionChannel. There (in OnMessageReceivedAsync() method) it is correctly detected and the Client.DisconnectAsync() gets called. 

Debugging showed that the code to cleanup the channels and the dispose of the TCP connection and stream object gets stuck in the line  258 of ChromeCastClient.Dispose(bool ...) method. I think what happens here is that the "**await ReceiveTcs.Task;**" never finalizes, because the Message Handlers of the channels get called **by the receive task itself,** and we hit a classical application deadlock here!

Note that calling the  Client.DisconnectAsync() method from 'the outside' (e.g. from GUI task) does work correctly and terminates the receive loop by exception, when the stream gets disposed.

I could fix this behavior by re-writing the message handler of the ConnectionChannel with the following:
```
public async override Task OnMessageReceivedAsync(IMessage message) {
            if (message is CloseMessage) {
                _ = Task.Run(() => {
                    Client.DisconnectAsync();
                });
            }
            await base.OnMessageReceivedAsync(message);
        }

```
By firing the DisconnectAsync from a new spawned Task, without blocking ('waiting for await callback') the receive loop is able to continue and finally the registered Disconnect event handlers get executed (out of the Task created here). Not sure if this solution is really waterproofed for all cases (e.g. if there are open Request Threads awaiting answers...) but I have no idea here how we could proof this with a unit test because we would have to simulate a device sending messages ....

**Comment by RobertK66** [comment]
Hi,
I did some more analyses/debugging in trying to wrap my head around this "async await stuff" ;-) ...

I found that it is not needed to 'spawn a new Task' with Task.Run(). What also solves the deadlock situation is that the Client.DisconnectAsync() gets called WITHOUT the 'await' keyword:
```
public async override Task OnMessageReceivedAsync(IMessage message) {
            if (message is CloseMessage) {
                _ =  Client.DisconnectAsync();
            }
            await base.OnMessageReceivedAsync(message);
        }

```

I think the reason for the deadlock is - as described above - that all this code runs in the receiver loop. This is the anonymous async task object created in Line 122 of the ChromeCasteClient.Receive function. My understanding of this async stuff is that a single async Task is either running some code lines with an assigned thread from the thread pool, or it is at some 'await' position of its code sequence waiting for a result (Future<x>) being generated/finalized by some other means. In case of a sequence of async await calls, the code running 'behind' the async call always gets a ('logical new') thread assigned for execution and if somewhere in that call hierarchy a result gets generated (with or without help of other background tasks) the 'await points' get this result passed -> fetch a new thread from the thread pool and continue to execute the lines after the await line.

Only by "ignoring" the not ready state of the "Future<x>" Task object returned by the async call, we can have both code paths continue to execute in parallel with 2 assigned working threads .

But in this original code here the DisconnectAsync() waits for the receive loop to be terminated, but the code to achieve this (failing the loop with an exception on the disposed socket connection) is never assigned a working thread because the receive loop itself waits for the OnMessageReceivedAsync() to finalize its (void) results ....

**Comment by Tapanila** [comment]
Will try to find some time testing this and write a fix for it

**Comment by RobertK66** [comment]
I can confirm this is fixed by Release 1.1.0 now - bits already downloaded from nuget ;-)
